File: frontend/src/models/file.ts
import { parseISO } from "date-fns";
import Folder from "./folder";

export interface FileResponse {
  ID: number;
  CreatedAt: string;
  UpdatedAt: string;
  DeletedAt: string | null;
  UserID: number;
  FolderID: number;
  FileName: string;
  FileCode: string;
  FileSize: number;
  FileType: string;  
  IsFavorite: boolean;
  Folder: Folder | null;
}

export class CloudChestFile {
  ID: number;
  CreatedAt: Date;
  UpdatedAt: Date;
  DeletedAt: Date | null;
  UserID: number;
  FolderID: number;
  FileName: string;
  FileCode: string;
  FileSize: number;
  FileType: string;
  IsFavorite: boolean;

  constructor(object: FileResponse) {
    this.ID = object.ID;
    this.CreatedAt = parseISO(object.CreatedAt);
    this.UpdatedAt = parseISO(object.UpdatedAt);
    this.DeletedAt = object.DeletedAt ? parseISO(object.DeletedAt) : null;
    this.UserID = object.UserID;
    this.FolderID = object.FolderID;
    this.FileName = object.FileName;
    this.FileCode = object.FileCode;
    this.FileSize = object.FileSize;
    this.FileType = object.FileType;
    this.IsFavorite = object.IsFavorite;
  }
}

export function parseFileResponse(response: FileResponse): CloudChestFile {
  return new CloudChestFile(response);
}

File: backend/internal/models/file.go
package models

import "gorm.io/gorm"

type FileUpdateBody struct {
	FileName   string `validate:"required" json:"file_name"`
	IsFavorite bool   `validate:"boolean" json:"is_favorite"`
	Restore    bool   `validate:"boolean" json:"is_restore"`
}

type FilePatchBody struct {
	FileName   string `json:"file_name"`
	FolderCode string `validate:"ascii" json:"folder_code"`
	IsFavorite bool   `validate:"boolean" json:"is_favorite"`
	Restore    bool   `validate:"boolean" json:"is_restore"`
}

type File struct {
	gorm.Model
	UserID        uint       `gorm:"not null"`
	FolderID      uint       `gorm:"not null"`
	FileName      string     `gorm:"type:varchar(255);not null"`
	FileCode      string     `gorm:"type:char(36);not null"`
	FileSize      uint       `gorm:"not null"`
	FileType      string     `gorm:"type:varchar(100);not null"`
	IsFavorite    bool       `gorm:"not null;default:0"`
	IsPreviewable bool       `gorm:"not null;default:0"`
	Folder        *Folder    `gorm:"foreignKey:FolderID"`
	Thumbnail     *Thumbnail `gorm:"foreignKey:FileID;constraint:OnDelete:CASCADE;"`
}
